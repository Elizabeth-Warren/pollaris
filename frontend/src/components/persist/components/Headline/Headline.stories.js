import React from 'react';
import styled from 'styled-components';
import { boolean, select, text } from '@storybook/addon-knobs';
import Headline, { HeadlineSizes } from './index';
import theme from '../../theme';

export default {
  title: 'Components/Headline',
  component: Headline,
  parameters: { jest: ['Headline'] },
};

export const withKnobs = () => {
  const tag = text('Tag', 'h1');
  const headlineText = text('Text', 'The quick brown fox jumps over the lazy dog.');
  const size = select('Size', Object.values(HeadlineSizes), Headline.defaultProps.size);
  const allCaps = boolean('All caps', Headline.defaultProps.allCaps);
  const breakpoint = text('Breakpoint', Headline.defaultProps.breakpoint);
  const color = select('Color', Object.keys(theme.colors), Headline.defaultProps.color);
  return (
    <div>
      <Headline
        as={tag}
        size={size}
        allCaps={allCaps}
        breakpoint={breakpoint}
        color={color}
      >
        {headlineText}
      </Headline>
    </div>
  );
};

export const gallery = () => {
  const headlineText = text('Text', 'The quick brown fox jumps over the lazy dog.');
  return (
    <div>
      <p>
        Note that the font size for the XL and LG headlines changes depending
        between mobile and tablet breakpoints.
      </p>
      {Object.entries(HeadlineSizes).reverse().map(([name, value], i) => (
        <Headline as={`h${i + 1}`} size={value} key={name}>
          {`${name}: `}
          {headlineText}
        </Headline>
      ))}
    </div>
  );
};

export const withStyledComponent = () => {
  const headlineText = text('Text', 'The quick brown fox jumps over the lazy dog.');
  const size = select('Size', Object.values(HeadlineSizes), Headline.defaultProps.size);
  const allCaps = boolean('All caps', Headline.defaultProps.allCaps);
  const breakpoint = text('Breakpoint', Headline.defaultProps.breakpoint);
  const StyledHeadline = styled.div`
    color: ${theme.colors.red};
  `;
  return (
    <div>
      <Headline
        as={StyledHeadline}
        size={size}
        allCaps={allCaps}
        breakpoint={breakpoint}
      >
        {headlineText}
      </Headline>
    </div>
  );
};
