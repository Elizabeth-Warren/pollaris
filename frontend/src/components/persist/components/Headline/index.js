import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import theme from '../../theme';

export const HeadlineSizes = {
  XS: 'xs',
  SM: 'sm',
  MD: 'md',
  LG: 'lg',
  XL: 'xl',
};

const HeadlineWrapper = styled.div`
  color: ${({ textColor }) => theme.colors[textColor] || 'inherit'};
  font-family: ${theme.fontFamily.compressed};
  font-size: ${({ size }) => {
    switch (size) {
      case HeadlineSizes.XL:
        return theme.fontSize['4xl'];
      case HeadlineSizes.MD:
        return theme.fontSize['2xl'];
      case HeadlineSizes.SM:
        return theme.fontSize.xl;
      case HeadlineSizes.XS:
        return theme.fontSize.lg;
      case HeadlineSizes.LG:
      default:
        return theme.fontSize['3xl'];
    }
  }};
  font-weight: bold;
  letter-spacing: ${({ size }) => {
    switch (size) {
      case HeadlineSizes.XL:
        return '-0.38px';
      default:
        return '0';
    }
  }};
  line-height: ${theme.leading.sub};
  text-transform: ${({ allCaps }) => (allCaps ? 'uppercase' : 'none')};

  @media (min-width: ${({ breakpoint }) => breakpoint}) {
    font-size: ${({ size }) => {
    switch (size) {
      case HeadlineSizes.XL:
        return theme.fontSize['5xl'];
      case HeadlineSizes.MD:
        return theme.fontSize['2xl'];
      case HeadlineSizes.SM:
        return theme.fontSize.xl;
      case HeadlineSizes.XS:
        return theme.fontSize.lg;
      case HeadlineSizes.LG:
      default:
        return theme.fontSize['4xl'];
    }
  }};
    letter-spacing: ${({ size }) => {
    switch (size) {
      case HeadlineSizes.XL:
        return '-0.9px';
      case HeadlineSizes.LG:
        return '-0.38px';
      default:
        return '0';
    }
  }}
  }
`;

const Headline = ({
  children, size, as, breakpoint, allCaps, color, ...other
}) => (
  <HeadlineWrapper
    allCaps={allCaps}
    as={as}
    breakpoint={breakpoint}
    size={size}
    textColor={color}
    {...other}
  >
    {children}
  </HeadlineWrapper>
);

Headline.propTypes = {
  /** Child text or elements */
  children: PropTypes.node.isRequired,
  /** HTML tag name or React element. */
  as: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
  /** Size of the headline */
  size: PropTypes.oneOf(Object.values(HeadlineSizes)),
  /** Color of the headline */
  color: PropTypes.oneOf(['inherit', ...Object.keys(theme.colors)]),
  /**
   * Screen width at or above which the headline takes on a larger font size.
   * Should be an em, rem, or px value.
   */
  breakpoint: PropTypes.string,
  /** Whether the headline should be in all caps. */
  allCaps: PropTypes.bool,
};

Headline.defaultProps = {
  size: HeadlineSizes.LG,
  breakpoint: theme.screens.md,
  allCaps: true,
  color: 'inherit',
};

export default Headline;
