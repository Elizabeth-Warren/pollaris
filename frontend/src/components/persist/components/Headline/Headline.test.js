import React from 'react';
import renderer from 'react-test-renderer';
import Headline, { HeadlineSizes } from './index';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <div>
        {Object.entries(HeadlineSizes).reverse().map(([name, value]) => (
          <Headline as="h1" size={value} key={name}>
            {`${name}: `}
            Hello world
          </Headline>
        ))}
      </div>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
