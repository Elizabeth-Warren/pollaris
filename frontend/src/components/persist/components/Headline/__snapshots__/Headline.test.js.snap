// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  color: inherit;
  font-family: 'Ringside Compressed A','Ringside Compressed B','Helvetica',sans-serif;
  font-size: 72px;
  font-weight: bold;
  -webkit-letter-spacing: -0.38px;
  -moz-letter-spacing: -0.38px;
  -ms-letter-spacing: -0.38px;
  letter-spacing: -0.38px;
  line-height: 0.9;
  text-transform: uppercase;
}

.c1 {
  color: inherit;
  font-family: 'Ringside Compressed A','Ringside Compressed B','Helvetica',sans-serif;
  font-size: 40px;
  font-weight: bold;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 0.9;
  text-transform: uppercase;
}

.c2 {
  color: inherit;
  font-family: 'Ringside Compressed A','Ringside Compressed B','Helvetica',sans-serif;
  font-size: 36px;
  font-weight: bold;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 0.9;
  text-transform: uppercase;
}

.c3 {
  color: inherit;
  font-family: 'Ringside Compressed A','Ringside Compressed B','Helvetica',sans-serif;
  font-size: 26px;
  font-weight: bold;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 0.9;
  text-transform: uppercase;
}

.c4 {
  color: inherit;
  font-family: 'Ringside Compressed A','Ringside Compressed B','Helvetica',sans-serif;
  font-size: 22px;
  font-weight: bold;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 0.9;
  text-transform: uppercase;
}

@media (min-width:768px) {
  .c0 {
    font-size: 110px;
    -webkit-letter-spacing: -0.9px;
    -moz-letter-spacing: -0.9px;
    -ms-letter-spacing: -0.9px;
    letter-spacing: -0.9px;
  }
}

@media (min-width:768px) {
  .c1 {
    font-size: 72px;
    -webkit-letter-spacing: -0.38px;
    -moz-letter-spacing: -0.38px;
    -ms-letter-spacing: -0.38px;
    letter-spacing: -0.38px;
  }
}

@media (min-width:768px) {
  .c2 {
    font-size: 36px;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 26px;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

@media (min-width:768px) {
  .c4 {
    font-size: 22px;
    -webkit-letter-spacing: 0;
    -moz-letter-spacing: 0;
    -ms-letter-spacing: 0;
    letter-spacing: 0;
  }
}

<div>
  <h1
    className="c0"
    size="xl"
  >
    XL: 
    Hello world
  </h1>
  <h1
    className="c1"
    size="lg"
  >
    LG: 
    Hello world
  </h1>
  <h1
    className="c2"
    size="md"
  >
    MD: 
    Hello world
  </h1>
  <h1
    className="c3"
    size="sm"
  >
    SM: 
    Hello world
  </h1>
  <h1
    className="c4"
    size="xs"
  >
    XS: 
    Hello world
  </h1>
</div>
`;
